modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/tests/PublisherIntegrationTestBase.cs

```diff
@@ -0,0 +1,196 @@
+    using Microsoft.Azure.IIoT.Modules.OpcUa.Publisher.Agent;
+    using Microsoft.Azure.IIoT.Modules.OpcUa.Publisher.Controller;
+    using Microsoft.Azure.IIoT.Modules.OpcUa.Publisher.Runtime;
+    using Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Engine;
+    using Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Models;
+    using Microsoft.Azure.IIoT.OpcUa.Protocol;
+    using Microsoft.Azure.IIoT.OpcUa.Protocol.Services;
+    using Microsoft.Azure.IIoT.Serializers;
+    using Microsoft.Azure.IIoT.Utils;
+    using Microsoft.Extensions.Configuration;
+    using Moq;
+    using Opc.Ua;
+    using Serilog;
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.IO;    
+    using System.Text;
+    using System.Text.Json;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using static Microsoft.Azure.IIoT.Hub.Mock.IoTHubServices;
+
+    /// <summary>
+    /// Base class for integration testing, it connects to the server, runs publisher and injects mocked IoTHub services.
+    /// </summary>
+    public class PublisherIntegrationTestBase {
+        /// <summary>
+        /// Whether the module is running.
+        /// </summary>
+        public BlockingCollection<EventMessage> Events { get; set; } = new BlockingCollection<EventMessage>();
+
+        public PublisherIntegrationTestBase() {
+            // This is a fake but correctly formatted connection string.
+            var connectionString = $"HostName=dummy.azure-devices.net; DeviceId={Utils.GetHostName()};SharedAccessKeyName=iothubowner;SharedAccessKey=aXRpc25vdGFuYWNjZXNza2V5";
+            var config = connectionString.ToIoTHubConfig();
+
+            _typedConnectionString = ConnectionString.Parse(config.IoTHubConnString);
+            _exit = new TaskCompletionSource<bool>();            
+        }
+
+        protected async Task<List<JsonDocument>> ProcessMessages(string publishedNodesFile) {
+            // publishedNodesFile points to the local server on the same machine and port currently as tests are run one at a time (not parallel) it does not need randomly generated port numbers.
+            _ = Task.Run(() => HostPublisherAsync(Mock.Of<ILogger>(), publishedNodesFile));
+
+            while (Events.Count == 0) {
+                await Task.Delay(500);
+            }
+
+            Exit();
+
+            var messages = new List<JsonDocument>();
+            foreach (var evt in Events) {
+                messages.Add(JsonDocument.Parse(Encoding.UTF8.GetString(evt.DisposedMessage)));
+            }
+
+            return messages;
+        }
+
+        /// <summary>
+        /// Setup publishing from sample server.
+        /// </summary>
+        protected async Task HostPublisherAsync(ILogger logger, string publishedNodesFilePath) {
+            try {
+                var config = _typedConnectionString.ToIoTHubConfig();
+
+                var arguments = new List<string>();
```

Comment:
It would be nice to have a way to set these for each test. For example, set PubSub, reversible encoding, etc. It could be an improvement later if you would prefer.